### ISSUE TITLE
Feature: Lifecycle evens for service implementations

### ISSUE THREAD

**Issue by rmichela** [grpc/grpc-java#3449]
I'd like my service implementations to be notified of server lifecycle events for server start and shutdown. These operations could be used to warm caches, persist data, or close files.

The least intrusive implementation would be an interface a service implementer would opt into for receiving these events. Ex:
```java
public interface ServiceLifecycle {
  // Called when Server.start() is called
  void onStart(...);
 
  // Called when Server.shutdown() or Server.shutdownNow() is called
  void onShutdown(boolean hard, ...);
}
```

My initial thought is that `onStart()` would block `Server.start()`, so that the server wouldn't be ready until all services had successfully started. `onShutdown()` is trickier. It needs to be called only after the server has drained all requests.

**Comment by ejona86** [comment]
During shutdown we already know when all requests have been drained, as this is when we transition to terminated. So that's feasible.

We used that sort of interface for Reflection, but someone noticed (I thought they were going to file an issue, but I don't see it...) that using an interceptor can break the interface-detection.

That interface would mean a service could only be used in one server. That's probably fine much of the time, but would maybe encourage people to want a server to be able to natively listen on multiple ports.

At this point we've expected people to do this sort of thing outside of gRPC. We'd need to think of implications doing it within, for example, error handling.

I would sort of expect people to find that interface convenient, as long as they aren't in some DI world that normally orchestrates startup/shutdown.

**Comment by rmichela** [comment]
I looked at how service lifecycle events could be implemented _on top_ of gRPC using an interceptor, rather than from within gRPC as a code change. The advantage would be less code in gRPC and an opt-in feature for those who are interested in it.

What I discovered is that there is no way to get from the `MethodDescriptor` returned by `ServerCall.getMethodDescriptor()` available in `ServerInterceptor.interceptCall()` to the actual class and method implementing the requested operation. The generated `MethodDescriptor` constants only contain information about the proto, with nothing about the binding itself.

Is there some way we can link a `MethodDescriptor` to the actual `java.lang.reflect.Method` that the descriptor is bound to? If we did this, I could implement this use case entirely outside of gRPC-java.

**Comment by ejona86** [comment]
#3469 is the issue with the existing `InternalNotifyOnBuild` used by reflection.

The [`ServerMethodDefinition`](https://grpc.io/grpc-java/javadoc/io/grpc/ServerMethodDefinition.html) is what specifies how to run the method. But the `ServerCallHandler` will be re-created when using interception.

We could add creation+destruction fields to [`ServerServiceDefinition`](https://grpc.io/grpc-java/javadoc/io/grpc/ServerServiceDefinition.html), with maybe some help from the generated code (like an instanceof check). @ericgribkoff, that might be a solution to `InternalNotifyOnBuild`

**Comment by jhaber** [comment]
Just wanted to mention some use-cases we have that this would be useful for. We run a separate netty server on a different port for admin-related stuff (inspired by dropwizard's admin servlet) and being able to shut this down when the grpc server shuts down would be really nice. In a similar vein, we have access logging that uses a background thread and would also like to stop this thread when the server shuts down.

In practice, our servers usually live for the life of the process, so we can usually get by with shutdown hooks. But for testing we might cycle through many servers rapidly to test different configurations and this ends up leaking various resources. So it's mainly a nice-to-have that would simplify some of our code.

**Comment by carl-mastrangelo** [comment]
@jhaber and @rmichela 

The gRPC Java team talked about this last week and there wasn't a clean solution.  Services have a life cycle that is different than a server, making it awkward to connect the two.  For example, a service can be shared by multiple servers. 

I was thinking of a solution here where there is a concept of a "Bound" services.  It would be a service tied to a particular server (or possible a surrogate object that represents the relation) which could accept lifecycle events.   Would this solve either of your use cases?

**Comment by jhaber** [comment]
Sorry for the noise, I think I misunderstood the issue title/description. I'm looking for lifecycle hooks for `io.grpc.Server`, not the individual services (since the resources I'm describing are tied to the server lifecycle). It seems like decorating the Server instance and intercepting calls gets me part of the way there, but I'm not sure if I'm missing something more obvious. I can also open a separate issue if you don't want to clutter this one up.

**Comment by ejona86** [comment]
@jhaber, for your case decorating the Server would be mostly-fine. You shouldn't need to intercept calls. Instead, you just need to intercept a call to shutdown/shutdownNow/awaitTermination.

**Comment by jhaber** [comment]
Makes sense, in case it's helpful for anyone else I put my fast-and-dirty approach in a [gist](https://gist.github.com/jhaber/1871b4ab87bb43bf86fa4cb580f88827). It's not perfect but seems to work well-enough for my use-case

**Comment by carl-mastrangelo** [comment]
This issue came up for me recently for Health status, and I like the idea in https://github.com/grpc/grpc-java/issues/3449#issuecomment-329938212

I see a couple lifecycle events being useful: prebuild, postbuild, prestart, poststart, preshutdown, postshutdown.    These would have to be optional, since custom Servers won't know about them.

**Comment by ejona86** [comment]
@carl-mastrangelo, it's not clear this would fix the health status issue, since 1) shutdown is too late of a notification, 2) health can't really change its state if it is being shared between two services.
